### ISSUE TITLE
MsPacManTask with HyperNEAT

### ISSUE THREAD

**Issue by schrum2** [schrum2/MM-NEATv2#105]
Thoroughly comment MsPacManTask and them make it implement the HyperNEATTask. This should probably wait until we have worked out the kinks with HyperNEAT in the pred/prey task though.

**Comment by leg2015** [comment]
see #29

**Comment by schrum2** [comment]
Here is how we will use HyperNEAT in Ms. Pac-Man. 

There will be four separate input substrates: 
1) One that senses regular pills
2) One that senses power pills
3) One that senses threat ghosts
4) One that senses edible ghosts

Each of these input grids needs to be big enough to cover the whole screen, even though certain input neurons will never be activated due to things like walls (the power pill substrate will be especially empty). We may look for ways to optimize this later and not try to make likes from neurons that can never be activated, but for now don't worry about it.

For now we'll do one processing layer, but we may make this a parameter later for more flexibility.

Now for the outputs:
The output grid will also be big enough to match the whole maze. The Ms. Pac-Man code has a function that will make her travel on the shortest path to any designated location in the maze. So, what we will do is overlay the output substrate on the maze and query all locations that contain valid locations (not walls). Whichever location has the highest activation is the one that Ms. Pac-Man will plot a course to.

I have other ideas for the output layer as well, but I think this one is the coolest, and since we have your substrate monitoring code, we should have a pretty easy time verifying that everything works as you develop it.

**Comment by schrum2** [comment]
Decided to share this task with Gabby. You can either work on it together, or whichever one of you gets to it first can take it on.

**Comment by leg2015** [comment]
Implemented hyperNEATTask in MsPacManTask, still need to add other hyperNEAT harnesses though

**Comment by leg2015** [comment]
Added needed method in NNHyperNEATPacManController, however still need to get the mediator all figured out

**Comment by leg2015** [comment]
So got all the basic code to implement hyperNEAT to ms PacMan set up, however the substrates produced are so huge that the code can't handle it(I added a println statement for every time a link is created and it creates SO many links). Will need to figure out a way to make the network smaller somehow if we want to be able to move forward in this domain...

**Comment by schrum2** [comment]
Keep removing pieces until we get something that at least runs. For starters, you could make the processing layer much smaller. We've seen in some Tetris runs how portions of the processing layer get ignored ... making it the same size as the input layers is just a rule of thumb. Try reducing the processing layer to 10 by 10, or even 5 by 5 if that doesn't work.

We could also drop the processing layer all-together and have the inputs connect directly to the output layer.

If even this doesn't work, then simply start removing input substrates until the code executes. We need to get an idea of how many links can be supported, and then we can consider ways of reconfiguring the substrates.

**Comment by leg2015** [comment]
So I finally got the network to construct if I cut out all but the msPacMan substrate and make the process substrate either nonexistent or really small. However, I believe there may be something a bit off with the outputs that are returned when process is called. Will have to investigate further, first by turning on watch and monitorSubstrates to see what happens

**Comment by leg2015** [comment]
Still not working yet, but there is a good amount of space between pills that we can compress to make then networks smaller and more efficient. Similar to Gabby's smudging

**Comment by schrum2** [comment]
I think most of the scaling problems can be fixed by changing the NNHyperNEATPacManController.getOutputIndexFromNodeCoord method, and also the MS_PAC_MAN_SUBSTRATE_WIDTH and MS_PAC_MAN_SUBSTRATE_HEIGHT. The trick is to make every 5x5 box in the original input space correspond to a single grid cell in the substrate ... I think this is the appropriate amount to scale by, but we'll need to play around with it a lot to be sure.

Basically, the empty space between pills in the actual display should not be present in the substrate ... adjacent pills should activate adjacent substrate grid cells.

**Comment by schrum2** [comment]
I made a lot of progress on this. All input substrates are smaller, and all of them work. I was able to make the processing layer slightly bigger, but not too big. The visual mode at least allows all components to display. The next step is to see if there is any chance of performing evolution within a reasonable time frame with this setup.

**Comment by leg2015** [comment]
MsPacMan is running with hyperNEAT, albeit not very well. Changed output substrate to query every position in the game space to see whether ms pacman should move towards there or not

**Comment by leg2015** [comment]
Got the nodes that aren't part of the maze to not link up. Works except for in visual without step-by-step. The reason is because there is a time limit to complete a move and the time it takes to set up the substrate visualizations is running over this time limit and crashing the program

**Comment by leg2015** [comment]
Added parameters that allow for separate threat and edible substrates. Also added code that creates a full screen power pill substrate. Also added option for there to be a larger processing substrate. These changes seem to run but there's still the odd error that we were getting with the substrate visualizations from earlier. Will need to investigate further...

**Comment by leg2015** [comment]
I think it's safe to call this closed since we have official runs of pacMan going.
